src/cargo/ops/cargo_generate_lockfile.rs

```diff
@@ -161,7 +164,49 @@ pub fn update_lockfile(ws: &Workspace<'_>, opts: &UpdateOptions<'_>) -> CargoRes
     let print_change = |status: &str, msg: String, color: &Style| {
         opts.config.shell().status_with_color(status, msg, color)
     };
-    for (removed, added) in compare_dependency_graphs(&previous_resolve, &resolve) {
+    let mut unchanged_behind = 0;
+    for (removed, added, unchanged) in compare_dependency_graphs(&previous_resolve, &resolve) {
+        let highest_present = [added.iter().rev().next(), unchanged.iter().rev().next()]
+            .into_iter()
+            .flatten()
+            .max_by_key(|s| s.version());
+        let latest = if let Some(present) = highest_present.filter(|p| p.source_id().is_registry())
```

Comment:
Maybe use a `filter_map()` for this too, to avoid the dangling `else { None }` branch at the end? Could use early return.

Also given that `latest` is prefiltered to be distinct from the current actual, I find the name not that obvious. Maybe `newer`?

Reply:
In my experience, the right balance of combiantors vs procedural logic is a matter of whether its bookkeeping and size.  In this case, this expression is too large in my opinion to shove in an iterator combinator.